builder/optimizations.cc

```diff
@@ -1689,4 +1700,300 @@ stmt canonicalize_nodes(const stmt& s) {
   return node_canonicalizer().mutate(s);
 }
 
+namespace {
+
+class task_parallelizer : public node_mutator {
+  std::set<var> consumed, produced;
+  bool barrier;
+  symbol_map<var> aliases;
+
+  var lookup_alias(var x) {
+    auto alias = aliases.lookup(x);
+    return alias ? *alias : x;
+  }
+
+  template <typename T>
+  static void remove_from_set(std::set<T>& set, const T& x) {
+    auto it = set.find(x);
+    if (it != set.end()) set.erase(it);
+  }
+
+  static bool should_async(const stmt& s) {
+    if (const block* b = s.as<block>()) {
+      for (const stmt& i : b->stmts) {
+        if (should_async(i)) return true;
+      }
+      return false;
+    } else if (s.as<check>()) {
+      return false;
+    } else {
+      return true;
+    }
+  }
+
+  // Here we are going to construct a DAG of stmts with their dependencies determining the edges.
+  struct stage {
+    stmt body;
+    std::set<var> consumed, produced;
+    std::set<stage*> consumers, producers;
+
+    // Fuse the producer `p` into this stage, the consumer, making `p` a no-op.
+    void fuse(stage* p) {
+      body = block::make(std::move(p->body), std::move(body));
+      for (stage* i : p->consumers) {
+        remove_from_set(i->producers, p);
+        i->producers.insert(this);
+      }
+      for (stage* i : p->producers) {
+        remove_from_set(i->consumers, p);
+        i->consumers.insert(this);
+      }
+      consumers.insert(p->consumers.begin(), p->consumers.end());
+      producers.insert(p->producers.begin(), p->producers.end());
+      consumed.insert(p->consumed.begin(), p->consumed.end());
+      produced.insert(p->produced.begin(), p->produced.end());
+
+      remove_from_set(consumers, this);
+      remove_from_set(consumers, p);
+      remove_from_set(producers, this);
+      remove_from_set(producers, p);
+    }
+  };
+
+  stage mutate_stage(const stmt& s) {
+    consumed.clear();
+    produced.clear();
+    stage result = {mutate(s)};
+    result.consumed = std::move(consumed);
+    result.produced = std::move(produced);
+    return result;
+  }
+
+public:
+  void visit(const block* b) override {
+    std::vector<stage> stages;
+    stages.reserve(b->stmts.size());
+
+    // Mutating the statements in the block, and determine which stages are producers for prior stages.
+    barrier = false;
```

Comment:
Same here, what if there is a block within block?

Reply:
It would be up to the block to propagate it. I deliberately didn't propagate this, but I realized that what we should do is propagate it until a loop, and then it should stop there.